### PR TITLE
Candidate implementation to speed up row extraction.

### LINKED ISSUE #100: Improve speed of extracting rows
With a global forecast dataset, I've the need to extract data with ~3,244.6k coordinate values (time x lat x lng). As of today, it takes about ~9 seconds to extract 1k rows. So: `9*3243.6/60/60` = ~8 hours per file.

While in streaming pipelines, data is written as soon as it's available, we'd ideally like all of the data to be processed within ~1 hour, so the first forecast is actionable.

### PR DESCRIPTION
**Pull request by alxmrs**
This PR attempts to improve the `ToBigQuery` Sink implementation by splitting up extraction into two parts. In the first part, groups of coordinate indexes are prepared. Second, rows are extracted from these coordinates. This aims to allow dataflow to parallelize and autoscale row extraction by asset.

The user now has a `coordinate_chunk_size` parameter they can tune. This allows them some control over how row extraction is parallelized. The autoscaling algorithm is controlled by CPU utilization of each [worker x task] pair. So, the user will need to balance increasing tasks to schedule more work (make chunks smaller) vs increase CPU utilization per task (make chunks bigger).

Fixes #100 (?).

## Remaining tasks:
- [x] Benchmark this branch against the current implementation (primarily on GCP)
  - [x] Investigate what the best value of `coordinate_chunk_size` is, set that as a default. The current value is chosen arbitrarily.
  - [x] To answer: is the overhead of opening datasets twice worth parallel execution?
- [x] Update documentation with the new arguments
- [x] Include additional performance improvements, as documented in #100
  - [x] In addition, investigate finding fast ways to get just coordinates from the dataset for `prepare_coordinates`. Do we need to implement #75 sooner?
- [ ] Add a section to the documentation about tuning extraction performance.

### CODE REVIEW

**Comment by alxmrs** [comment]
Another possible improvement to this PR: The logging for `get_coordinates()` will probably be weird as it is now. The log statements for that function could be removed or brought into `prepare_coordinates()`.

File: weather_mv/loader_pipeline/bq.py [hunk 1]
```diff
@@ -103,28 +108,38 @@ def expand(self, paths):
...
                 paths
-                | 'ExtractRows' >> beam.FlatMap(
+                | 'PrepareCoordinates' >>
+                beam.FlatMap(
+                    prepare_coordinates,
+                    coordinate_chunk_size=self.coordinate_chunk_size,
+                    area=self.area,
+                    open_dataset_kwargs=self.xarray_open_dataset_kwargs,
+                )
+                | beam.Reshuffle()
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/bq.py:118]: @ksic8 I believe that this is the best place for a reshuffle, as we discussed in today's standup.

**Comment by ksic8** [comment]
PFB link of excel containing results of test-cases that showcases improvised performance,
[GH #100 (Improve speed of extracting rows) Results v1.0](https://docs.google.com/spreadsheets/d/1yfJ7ne5y4u15HTpliwRZc1GOBovTUhl3N07zo0_4q48/edit?usp=sharing)

**Comment by alxmrs** [comment]
I used this micro-benchmark test that Pramod wrote to compare different implementations of this routine:
https://github.com/google/weather-tools/blob/8d88e67f8964fa357305ed2800bd66bd05fc5367/weather_mv/loader_pipeline/bq_test.py#L383

Here's a comparison of running cProfile on this test with different implementations:
- [a754a80](https://github.com/google/weather-tools/pull/146/commits/a754a8048961ba81874286d2265a65933fbc2523) - 40.28s
- [8f16f09](https://github.com/google/weather-tools/pull/146/commits/8f16f09bd9450b871d21b3c17813b4c1d136e3aa) - 31.29s

@ksic8 Can you run an e2e benchmark on the latest change to see if it's even faster?

**Comment by ksic8** [comment]
@alxmrs revised-code & documentation changes pushed into github. Bench-marking results added in [google-sheet](https://docs.google.com/spreadsheets/d/1wiugaiTF7UIySO6y8wtSZUAyMRyaEXkuPlHXWJdFl-U/edit#gid=0).

File: weather_mv/loader_pipeline/bq.py [hunk 2]
```diff
@@ -190,14 +221,12 @@ def extract_rows(uri: str, *,
...
-        if area:
-            n, w, s, e = area
-            data_ds = data_ds.sel(latitude=slice(n, s), longitude=slice(w, e))
-            logger.info(f'Data filtered by area, size: {data_ds.nbytes}')
-
         first_ts_raw = data_ds.time[0].values if data_ds.time.size > 1 else data_ds.time.values
         first_time_step = to_json_serializable_type(first_ts_raw)
 
+        # Applying serialization function over entire dataset
+        data_ds.map_blocks(to_json_serializable_type)
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_mv/loader_pipeline/bq.py:224]: @ksic8 To reiterate the above statement, this line doesn't work as intended, and probably isn't needed. 

I just stepped into the serialization function with the debugger. What happens in this `map_blocks` call is that the data frame reaches `np.isnan(value)` and this returns the datafame, which is interpreted as `True`, and returns `None`. 

However, `map_blocks` doesn't mutate the input in the first place, it returns the new dataset. Here, the dataset is being dropped. 

The solution is either to: 
1. Use the dictionary comprehension version of the code that I had in a previous commit
2. Remove this line and be happy that serialization is already happening, but wonder if at some time in the future, things will break.

[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: @alxmrs revised functionality committed for your review. Test-case details are available in [Link](https://docs.google.com/spreadsheets/d/1yfJ7ne5y4u15HTpliwRZc1GOBovTUhl3N07zo0_4q48/edit?usp=sharing). (Refer Sheet 2)

**Review by ksic8 — COMMENTED** [review]
@alxmrs revised code submitted for your review.

File: weather_mv/README.md [hunk 3]
```diff
@@ -160,3 +160,34 @@ weather-mv --uris "gs://your-bucket/*.nc" \
...
 ```
+
+## Private Network Configuration
+
+While running `weather-mv` pipeline in GCP, there is a possibility that you may receive following error -
+"Quotas were exceeded: IN_USE_ADDRESSES"
+
+This error occurs when GCP is trying to add new worker-instances and finds that, “Public IP” quota (assigned to your project) is exhausted.
+
+To solve this, it is recommended to use private IP while running your dataflow pipelines.
```
Comment (alxmrs) [inline weather_mv/README.md:171]:
```suggestion
To solve this, we recommended using private IP while running your dataflow pipelines.
```

**Review by alxmrs — COMMENTED** [review]
really minor copy-editing nit :)

File: weather_mv/loader_pipeline/bq.py [hunk 4]
```diff
@@ -16,19 +16,20 @@
 import datetime
 import logging
 import typing as t
+from pprint import pformat
+import pandas as pd
+import math
+import pytz
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Removed pytz dependency.

File: weather_mv/loader_pipeline/bq.py [hunk 5]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
-    # re-calculate import time for streaming extractions.
-    if not import_time:
-        import_time = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc).isoformat()
-
+def prepare_coordinates(uri: str, *,
+                        coordinate_chunk_size: int,
+                        area: t.Optional[t.List[int]] = None,
+                        import_time: t.Optional[str] = DEFAULT_IMPORT_TIME,
+                        open_dataset_kwargs: t.Optional[t.Dict] = None,
+                        variables: t.Optional[t.List[str]] = None) -> t.Iterator[t.Tuple[str, str, str, pd.DataFrame]]:
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 6]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
-            row[GEO_POINT_COLUMN] = fetch_geo_point(row['latitude'], row['longitude'])
-
-            # 'row' ends up looking like:
-            # {'latitude': 88.0, 'longitude': 2.0, 'time': '2015-01-01 06:00:00', 'd': -2.0187, 'cc': 0.007812,
-            #  'z': 50049.8, 'data_import_time': '2020-12-05 00:12:02.424573 UTC', ...}
-            beam.metrics.Metrics.counter('Success', 'ExtractRows').inc()
-            return row
-
-        yield from map(to_row, get_coordinates(data_ds, uri))
+        # add co-ordinates for 0-dimension file.
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Comments modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 7]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+        if (len(data_ds.to_array().shape) == 1):
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 8]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+                data_ds = data_ds.assign_coords({coord: [data_ds[coord].values]})
+
+        df = data_ds.to_dataframe().reset_index()
+
+        # Add un-indexed coordinates.
+        if variables:
+            indices = list(ds.coords.indexes.keys())
+            to_keep = variables + indices
+            to_drop = set(df.columns) - set(to_keep)
+            df.drop(to_drop, axis=1, inplace=True)
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Comments modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 9]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+        num_chunks = math.ceil(len(df) / coordinate_chunk_size)
+        for i in range(num_chunks):
+            chunk = df[i * coordinate_chunk_size:(i + 1) * coordinate_chunk_size]
+            yield uri, import_time, first_time_step, chunk
+
+
+def __convert_time(val) -> t.Any:
+    """Converts pandas Timestamp values to ISO format."""
+    if isinstance(val, pd.Timestamp):
+        return val.tz_localize(pytz.UTC).isoformat()
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 10]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+            to_drop = set(df.columns) - set(to_keep)
+            df.drop(to_drop, axis=1, inplace=True)
+
+        num_chunks = math.ceil(len(df) / coordinate_chunk_size)
+        for i in range(num_chunks):
+            chunk = df[i * coordinate_chunk_size:(i + 1) * coordinate_chunk_size]
+            yield uri, import_time, first_time_step, chunk
+
+
+def __convert_time(val) -> t.Any:
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 11]
```diff
@@ -103,13 +108,19 @@ def expand(self, paths):
...
                     import_time=self.import_time,
-                    open_dataset_kwargs=self.xarray_open_dataset_kwargs)
-            )
+                    open_dataset_kwargs=self.xarray_open_dataset_kwargs,
+                    variables=self.variables
+                )
+                | beam.Reshuffle()
+                | 'ExtractRows' >> beam.FlatMapTuple(
+                    extract_rows
+                )
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 12]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+
+def extract_rows(uri: str,
+                 import_time: str,
+                 first_time_step: str,
+                 rows: pd.DataFrame) -> t.Iterator[t.Dict]:
+    """Reads an asset and coordinates, then yields its rows as a mapping of column names to values."""
+    for _, row in rows.iterrows():
+        row = row.astype(object).where(pd.notnull(row), None)
+        row = row.to_dict()
+        row = {k: __convert_time(v) for k, v in row.items()}
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py]: Code modified as per review comment.

File: weather_mv/README.md (hunk 3 above)
Comment (ksic8) [reply weather_mv/README.md]: Remarks modified as per review comment.

File: weather_mv/loader_pipeline/bq.py [hunk 13]
```diff
@@ -175,58 +186,72 @@ def fetch_geo_point(lat: float, long: float) -> str:
...
+        return val
+
+
+def extract_rows(uri: str,
+                 import_time: str,
+                 first_time_step: str,
+                 rows: pd.DataFrame) -> t.Iterator[t.Dict]:
+    """Reads an asset and coordinates, then yields its rows as a mapping of column names to values."""
+    for _, row in rows.iterrows():
+        row = row.astype(object).where(pd.notnull(row), None)
```
[the remark this answers is not in the document]
Comment (ksic8) [reply weather_mv/loader_pipeline/bq.py:245]: @alxmrs We tested outcome against both code variants - processing per row (For-loop) vs. processing whole data-frame (Pandas.DataFrame.apply). And found that, processing-per-row results in better performance for our scenario.

[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_mv/loader_pipeline/bq.py:245]: Cool, thanks for the explanation!

**Comment by alxmrs** [comment]
According to this tool, I can't approve this CL since I was the original author. However, this CL LGTM – please resolve the merge conflicts, and then you can merge.

**Pull request merged**